clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -869,15 +883,34 @@ protected boolean isDynamicMember() {
     private class LeaveGroupResponseHandler extends CoordinatorResponseHandler<LeaveGroupResponse, Void> {
         @Override
         public void handle(LeaveGroupResponse leaveResponse, RequestFuture<Void> future) {
-            Errors error = leaveResponse.error();
+            final Errors error;
+            if (leaveResponse.data().memberResponseListV0().isEmpty()) {
+                error = getSingleLeaveError(leaveResponse);
+            } else {
+                error = getBatchLeaveError(leaveResponse, future);
+            }
             if (error == Errors.NONE) {
                 log.debug("LeaveGroup request returned successfully");
                 future.complete(null);
             } else {
-                log.debug("LeaveGroup request failed with error: {}", error.message());
+                log.error("LeaveGroup request failed with error: {}", error.message());
                 future.raise(error);
             }
         }
+
+        private Errors getSingleLeaveError(LeaveGroupResponse leaveResponse) {
+            return Errors.forCode(leaveResponse.data().errorCode());
+        }
+
+        private Errors getBatchLeaveError(LeaveGroupResponse leaveResponse,
+                                          RequestFuture<Void> future) {
+            List<MemberResponse> memberResponseListV0 = leaveResponse.data().memberResponseListV0();
+            if (memberResponseListV0.size() != 1) {
+                future.raise(new UnsupportedForMessageFormatException("The expected leave group response " +
```

Comment:
I think this could just be `IllegalStateException` or something like that. It doesn't really have anything to do with the message format.